src/qminer/qminer_core.cpp

```diff
@@ -5877,31 +5910,66 @@ void TIndex::IndexTextPos(const int& KeyId, const TStr& TextStr, const uint64& R
     IndexTextPos(KeyId, WordIdV, RecId);
 }
 
-void TIndex::IndexTextPos(const int& KeyId, const TUInt64V& WordIdV, const uint64& RecId) {
+void TIndex::UpdateTextPos(const int& KeyId, const TUInt64V& WordIdV, const uint64& RecId, void (TGix<TQmGixKey, TQmGixItemPos>::*UpdateMethod)(const TQmGixKey&, const TQmGixItemPos&)) {
     // we shouldn't modify read-only index
     QmAssertR(!IsReadOnly(), "Cannot edit read-only index!");
-    // aggregate by word
-    THash<TUInt64, TQmGixItemPos> WordIdPosH;
-    for (int WordIdN = 0; WordIdN < WordIdV.Len(); WordIdN++) {
-        const uint64 WordId = WordIdV[WordIdN];
-        // check if first time we see the word
-        if (!WordIdPosH.IsKey(WordId)) {
-            WordIdPosH.AddDat(WordId, TQmGixItemPos(RecId));
+    // for each word we want to build the TQmGixItemPos that is already sorted from beginning. For that reason we
```

Comment:
It is unfortunately not as simple. If the document will have more than 1024 words, this will not ensure that the values in TQmGixItemPos will be in correct order.  If we would have a single loop we would have to build for each word an arbitrary long vector of positions and then at the end of the loop sort them and insert in groups of 3 to gix. The current code, although complicated, ensures that this is done as efficiently as possible.